### PR TITLE
feat: add build info to metadata

### PR DESCRIPTION
**Pull request by itegulov**
Some of the metadata could be made more type safe (e.g. `compiler` should be `(Language, Version)` formatted with a whitespace inbetween), but I think it is fine as is for now.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -54,6 +54,19 @@ fn ensure_current_version<'de, D: Deserializer<'de>>(d: D) -> Result<String, D::
     Ok(unchecked)
 }
 
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Default)]
+pub struct BuildInfo {
+    /// The hash of the contract WASM code.
+    pub wasm_hash: String,
```
Comment (miraclx) [inline src/lib.rs]: Unsure how we're to represent this, we should probably just make it an `Option`, in the case that the ABI is embedded, otherwise this would be cyclic.

Comment (austinabell) [inline src/lib.rs]: Opinionated: This doesn't seem related to the build info at all, it's just the code hash. Wondering if it should just be in the root metadata?

**Review by austinabell — APPROVED** [review]
My comment is opinionated, so generally LGTM

File: src/lib.rs (hunk 1 above)
Comment (austinabell) [reply src/lib.rs]: interesting, that could be valid. I was under the assumption that build info in general would be omitted when embedding, but that is valid that one might want the other build info in the embedded ABI metadata

Comment (itegulov) [reply src/lib.rs]: Yes, my intuition was for the build info to be omitted entirely when embedding, but I guess an argument could be made that someone would want some of this info inside wasm.

Comment (itegulov) [reply src/lib.rs]: Hmm, fair, especially because it also resolved Mira's point above

**Review by miraclx — APPROVED** [review]
lgtm

File: src/lib.rs (hunk 1 above)
Comment (itegulov) [reply src/lib.rs]: Moved `wasm_hash` to metadata root, so build info can be embedded now if we wanted to.
